pkg/cloudprovider/provider/vsphere/helper.go

```diff
@@ -106,21 +106,48 @@ func CreateLinkClonedVm(vmName, vmImage, datacenter, clusterName, folder string,
 	if containerLinuxUserdata != "" {
 		userdataBase64 := base64.StdEncoding.EncodeToString([]byte(containerLinuxUserdata))
 
-		//Found out via `govc  vm.info -json containerlinux-template`
-		//Both `Key` and `Id` must be correct in order for this to work
-		//You will however not get an error from the API if its wrong
-		//Only the `Id` is part of the source OVA image, the `Key` gets generated by Vsphere
-		//and I have no idea if that generation is deterministic
-		vAppAconfig = &types.VmConfigSpec{Property: []types.VAppPropertySpec{
-			{
-				Info:            &types.VAppPropertyInfo{Key: 7, Id: "guestinfo.coreos.config.data.encoding", Value: "base64"},
-				ArrayUpdateSpec: types.ArrayUpdateSpec{Operation: types.ArrayUpdateOperationEdit},
-			},
-			{
-				Info:            &types.VAppPropertyInfo{Key: 4, Id: "guestinfo.coreos.config.data", Value: userdataBase64},
-				ArrayUpdateSpec: types.ArrayUpdateSpec{Operation: types.ArrayUpdateOperationEdit},
-			},
-		}}
+		// The properties describing userdata will already exist in the CoreOS VM template.
+		// In order to overwrite them, we need to specify their numeric Key values,
+		// which we'll extract from that template.
+		var mvm mo.VirtualMachine
+		err = templateVm.Properties(ctx, templateVm.Reference(), []string{"config", "config.vAppConfig", "config.vAppConfig.property"}, &mvm)
+		if err != nil {
+			return err
+		}
+
+		var property []types.VAppPropertySpec
+		info := mvm.Config.VAppConfig.GetVmConfigInfo()
+		properties := info.Property
+		for _, item := range properties {
+			switch item.Id {
+			case "guestinfo.coreos.config.data":
+				property = append(property, types.VAppPropertySpec{
+					ArrayUpdateSpec: types.ArrayUpdateSpec{
+						Operation: types.ArrayUpdateOperationEdit,
+					},
+					Info: &types.VAppPropertyInfo{
+						Key:   item.Key,
+						Id:    item.Id,
+						Value: userdataBase64,
+					},
+				})
+				break
+			case "guestinfo.coreos.config.data.encoding":
+				property = append(property, types.VAppPropertySpec{
+					ArrayUpdateSpec: types.ArrayUpdateSpec{
+						Operation: types.ArrayUpdateOperationEdit,
+					},
+					Info: &types.VAppPropertyInfo{
+						Key:   item.Key,
+						Id:    item.Id,
+						Value: "base64",
+					},
+				})
+				break
+			}
+		}
+
+		vAppAconfig = &types.VmConfigSpec{Property: property}
```

Comment:
oh, we are taking only the first encountered value.

Reply:
`property` was actually a slice. But as the naming of the vmware type is fairly bad, it's super confusing.
I renamed the var to make it more clear